### ISSUE TITLE
Live Web Preview

### ISSUE THREAD

**Issue by MattEllison** [Antaris/RazorEngine#330]
Hi I've been trying to review the documentation but not sure I understand enough to solve this problem. Basically i'm working on a templating project where I'm essentially recreating a mail merge process. The Razor engine works great except for one little issue I'm having.

Using ASP.NET I'm trying to generate a live preview of a template the user is working on. So every time the user makes a change on the text editor I send the HTML via ajax back to the server to process is and return it back with some sample data. So I want to run the new HTML data through the the Razor engine each time but considering the user is constantly making changes it seems it would require a new template key with each update. 

I could just use a web based template engine for this scenario but I plan to run some scheduled task on the server side in the future so sticking with the razor syntax seemed ideal.

I see from other comments its difficult to just reload a template key or remove and reload so this RazorEngine may not be the best tool for me to use in this scenario however I figured I would check first and see if someone has some suggestions.

I apologize for any technical ignorance I have but would appreciate comments or suggestions.

Thanks

**Comment by matthid** [comment]
Yes that would be a scenario where RazorEngine is not ideal because of the following reasons:

- RazorEngine is SLOW for repeatedly building CHANGED or DIFFERENT templates.
- This is by design, the advantage is that RazorEngine is *incredible* fast for REPEATEDLY building the SAME template because after all the overhead it is basically simply a method call.

If the live templating is an exception in your scenario (ie the finished template is used later more often) and you can live with longer waiting times on the live preview (and actually prefer faster times later) things become slightly complicated (for the live templating at least)...

Because we basically transform templates to assemblies and load them, the memory overhead is huge and you need to implement some kind of AppDomain recycling (which in itself is already complicated) to unload existing AppDomain's and create new ones (on every X change).

The good news are that depending on you ASP.NET scenario IIS might already do that for you.

Now to the RazorEngine (technical) side: Because deleting things from the cache is basically a memory leak (for the reasons specified above) we don't allow it in the default settings. We ship however all you need, please read http://antaris.github.io/RazorEngine/TemplateManager.html and tell me if something is unclear.

TL;DR Only use RazorEngine if it is more important to run the template fast later (after editing) and it is not important for the live template to be created "really fast".

Hope this helps.
